test/widgets/compose_box_test.dart

```diff
@@ -195,10 +206,163 @@ void main() {
+    Future<void> prepareComposeBoxWithNavigation(WidgetTester tester) async {
+      addTearDown(testBinding.reset);
+      await testBinding.globalStore.add(account, eg.initialSnapshot(
+        zulipFeatureLevel: account.zulipFeatureLevel));
+
+      await tester.pumpWidget(const ZulipApp());
+      await tester.pump();
+      final navigator = await ZulipApp.navigator;
+      unawaited(navigator.push(MaterialAccountWidgetRoute(
+        accountId: account.id, page: const ComposeBox(narrow: narrow))));
+      await tester.pumpAndSettle();
+
+      store = await testBinding.globalStore.perAccount(account.id);
+      connection = store.connection as FakeApiConnection;
+    }
+
+    testWidgets('navigating away sends "stopped typing" notice', (tester) async {
+      await prepareComposeBoxWithNavigation(tester);
+
+      await checkStartTyping(tester, narrow);
+
+      connection.prepare(json: {});
+      (await ZulipApp.navigator).pop();
+      await tester.pump(Duration.zero);
+      checkTypingRequest(TypingOp.stop, narrow);
+    });
+
+    testWidgets('for content input, unfocusing sends "stopped typing" notice '
+                'and refocusing does not send "started typing notice', (tester) async {
+      const narrow = ChannelNarrow(123);
+      final destinationNarrow = TopicNarrow(narrow.streamId, 'test topic');
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await tester.enterText(topicInputFinder, destinationNarrow.topic);
+      // Remove an irrelevant topic request.
+      check(connection.takeRequests()).single
+        ..method.equals('GET')
+        ..url.path.equals('/api/v1/users/me/123/topics');
+
+      connection.prepare(json: {});
+      await tester.enterText(contentInputFinder, 'hello world');
+      checkTypingRequest(TypingOp.start, destinationNarrow);
+
+      connection.prepare(json: {});
+      // Move focus to the topic input.
+      await tester.tap(topicInputFinder);
+      await tester.pump(Duration.zero);
+      checkTypingRequest(TypingOp.stop, destinationNarrow);
+
+      await tester.tap(contentInputFinder);
+      await tester.pump(Duration.zero);
+      // Refocusing on the content input triggers an update to
+      // [TextEditingController.selection].  Still, we expect no
+      // "typing started" request because the text remains the same.
+    });
+
+    testWidgets('unfocusing app sends "typing stopped" notice', (tester) async {
+      Future<void> setAppLifeCycleState(AppLifecycleState state) async {
+        // While this state lives on [ServicesBinding], testWidgets resets it
+        // for us when the test ends:
+        //   https://github.com/flutter/flutter/blob/c78c166e3ecf963ca29ed503e710fd3c71eda5c9/packages/flutter_test/lib/src/binding.dart#L1189
+        final ByteData? message = const StringCodec().encodeMessage(state.toString());
+        await tester.binding.defaultBinaryMessenger
+          .handlePlatformMessage('flutter/lifecycle', message, (_) {});
```

Comment:
Let's try to avoid reaching like this into the implementation details of how `didChangeAppLifecycleState` gets invoked.

See a similar case here:
https://github.com/zulip/zulip-flutter/pull/856#discussion_r1809705326

Reply:
That's pretty neat!